quinn-proto/src/tests/util.rs

```diff
@@ -333,17 +338,17 @@ impl TestEndpoint {
         loop {
             let mut endpoint_events: Vec<(ConnectionHandle, EndpointEvent)> = vec![];
             for (ch, conn) in self.connections.iter_mut() {
-                if self.timeout.map_or(false, |x| x <= now) {
-                    self.timeout = None;
-                    conn.handle_timeout(now);
-                }
-
                 for (_, mut events) in self.conn_events.drain() {
                     for event in events.drain(..) {
                         conn.handle_event(event);
                     }
                 }
 
+                if self.timeout.map_or(false, |x| x <= now) {
+                    self.timeout = None;
+                    conn.handle_timeout(now);
+                }
+
```

Comment:
What's the significance of relocating this?

Reply:
(And relocating code should go in a separate commit.)

Reply:
I don't exactly remember why this was necessary, but I just reverted it and observed that at least one test triggers an infinite loop (probably related to the timers that are being added in this PR). I'll extract the change into its own commit.

Reply:
Assuming we tweak the timer as mentioned [here](https://github.com/quinn-rs/quinn/pull/1553#discussion_r1251222791), I'm afraid requiring two unacked packets is too relaxed. Think of the following scenario:

1. We send a single ack-eliciting packet after a period of inactivity. Because the RTT timer is unset, the packet automatically gets an `IMMEDIATE_ACK` frame and the RTT timer is set.
2. We receive an ACK before the RTT timer elapses.
3. We send a new packet. Because the RTT timer is set, the packet does not get an `IMMEDIATE_ACK` frame, and the RTT timer isn't touched.
4. The RTT timer elapses, and sees that there is only one packet in flight. The timer is unset and no `IMMEDIATE_ACK` frame is sent, _which is not according to spec_.

Maybe we should track timeout-triggered `IMMEDIATE_ACK` frames after all? Not sure about the best way to do it, but maybe we could modify `PacketSpace::immediate_ack_pending` to be of type `Option<ImmediateAckSource>`:

```rust
enum ImmediateAckSource {
    Timeout,
    Other,
}
```

That way, when sending the packet, we can record its number if the source of the immediate ack was a timeout.

Reply:
Note also that tail loss probes will already be sent by the existing loss detection logic if we don't get an ACK within a constant factor of an RTT for a trailing flight.